pilot/pkg/proxy/envoy/v2/debug.go

```diff
@@ -214,8 +214,8 @@ func (s *DiscoveryServer) endpointShardz(w http.ResponseWriter, req *http.Reques
 	_ = req.ParseForm()
 	w.Header().Add("Content-Type", "application/json")
 	s.mutex.RLock()
+	defer s.mutex.RUnlock()
```

Comment:
I may be misunderstanding the flow here, but this seems off. The lock is intended to protect access to `s.EndpointShardsByService`, right? Holding it while writing the response to the wire seems bad.

Reply:
I completely agree, however that's the current implementation and this is just guarding against the unlock not happening because of a panic during the writing of the response.

Reply:
Ah, yeah, fair enough. Did this come from upstream or something, or is this our implementation? Haven't seen any mention of upstream issues, or whether this needs to be addressed in the 2.1 branch.

Anyway, would a `getEndpointShardsJSON()` method, similar to how the other bits were split out, be better? I just worry that this may introduce a new, different, possibility for a DOS.

Reply:
It's upstream code and nothing changes functionally, all this change does is remove the possibility of a DoS by ensuring the lock is always released regardless of how the function terminates.

Reply:
Actually I see what you mean and will update it